lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,918 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "resolve-labels.h"
+#include "../common/enum-set.h"
+#include "../parser/message.h"
+#include "../parser/parse-tree-visitor.h"
+#include <cctype>
+#include <cstdarg>
+#include <iostream>
+
+namespace Fortran::semantics {
+
+using namespace parser::literals;
+
+ENUM_CLASS(TargetStatementEnum, Do, Branch, Format)
+using LabeledStmtClassificationSet =
+    common::EnumSet<TargetStatementEnum, TargetStatementEnum_enumSize>;
+
+using IndexList = std::vector<std::pair<parser::CharBlock, parser::CharBlock>>;
+// A ProxyForScope is an integral proxy for a Fortran scope. This is required
+// because the parse tree does not actually have the scopes required.
+using ProxyForScope = unsigned;
+struct LabeledStatementInfoTuplePOD {
```

Comment:
I think `LabeledStatementInfo` would be a better name. Same with the next struct.

Reply:
Yeah, but my compiler was wringing its hands about these simple PODs. The extra garbage lets me find them a little easier. Deferring.